pkg/core/native/ledger.go

```diff
@@ -0,0 +1,230 @@
+	md = newMethodAndPrice(l.getTransactionHeight, 1000000, callflag.ReadStates)
+	l.AddMethod(md, desc)
+
+	desc = newDescriptor("getTransactionFromBlock", smartcontract.ArrayType,
+		manifest.NewParameter("blockIndexOrHash", smartcontract.ByteArrayType),
+		manifest.NewParameter("txIndex", smartcontract.IntegerType))
+	md = newMethodAndPrice(l.getTransactionFromBlock, 2000000, callflag.ReadStates)
+	l.AddMethod(md, desc)
+
+	return l
+}
+
+// Metadata implements Contract interface.
+func (l *Ledger) Metadata() *interop.ContractMD {
+	return &l.ContractMD
+}
+
+// Initialize implements Contract interface.
+func (l *Ledger) Initialize(ic *interop.Context) error {
+	return nil
+}
+
+// OnPersist implements Contract interface.
+func (l *Ledger) OnPersist(ic *interop.Context) error {
+	// Actual block/tx processing is done in Blockchain.storeBlock().
+	// Even though C# node add them to storage here, they're not
+	// accessible to smart contracts (see isTraceableBlock()), thus
+	// the end effect is the same.
+	return nil
+}
+
+// PostPersist implements Contract interface.
+func (l *Ledger) PostPersist(ic *interop.Context) error {
+	return nil // Actual block/tx processing is done in Blockchain.storeBlock().
+}
+
+// currentHash implements currentHash SC method.
+func (l *Ledger) currentHash(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.Make(ic.Chain.CurrentBlockHash().BytesBE())
+}
+
+// currentIndex implements currentIndex SC method.
+func (l *Ledger) currentIndex(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.Make(ic.Chain.BlockHeight())
+}
+
+// getBlock implements getBlock SC method.
+func (l *Ledger) getBlock(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	hash := getBlockHashFromItem(ic.Chain, params[0])
+	block, err := ic.Chain.GetBlock(hash)
+	if err != nil || !isTraceableBlock(ic.Chain, block.Index) {
+		return stackitem.Null{}
+	}
+	return BlockToStackItem(block)
+}
+
+// getTransaction returns transaction to the SC.
+func (l *Ledger) getTransaction(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	tx, h, err := getTransactionAndHeight(ic.DAO, params[0])
+	if err != nil || !isTraceableBlock(ic.Chain, h) {
+		return stackitem.Null{}
+	}
+	return TransactionToStackItem(tx)
+}
+
+// getTransactionHeight returns transaction height to the SC.
+func (l *Ledger) getTransactionHeight(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	_, h, err := getTransactionAndHeight(ic.DAO, params[0])
+	if err != nil || !isTraceableBlock(ic.Chain, h) {
+		return stackitem.Make(-1)
+	}
+	return stackitem.Make(h)
+}
+
+// getTransactionFromBlock returns transaction with the given index from the
+// block with height or hash specified.
+func (l *Ledger) getTransactionFromBlock(ic *interop.Context, params []stackitem.Item) stackitem.Item {
+	hash := getBlockHashFromItem(ic.Chain, params[0])
+	index, err := params[1].TryInteger()
```

Comment:
`toBigInt`? Or even `toUint32`, as `MaxTransactionsPerBlock` is less than `math.MaxUint16`.

Reply:
Fixed.